File: archive/src/main/kotlin/com/grappenmaker/aoc/year25/day10.py
import pyperclip
from z3 import Optimize, Int

machines = [([[3], [1, 3], [2], [2, 3], [0, 2], [0, 1]], [3, 5, 4, 7]),
([[0, 2, 3, 4], [2, 3], [0, 4], [0, 1, 2], [1, 2, 3, 4]], [7, 5, 12, 7, 2]),
([[0, 1, 2, 3, 4], [0, 3, 4], [0, 1, 2, 4, 5], [1, 2]], [10, 11, 11, 5, 10, 5])]

def solve(machine):
  solver = Optimize()
  buttons, jolts = machine
  buttons = [set(button) for button in buttons]

  button_vars = [Int(f'b{i}') for i in range(len(buttons))]
  for i in range(len(jolts)):
    lhs = sum(button_vars[j] for j in range(len(buttons)) if i in buttons[j])
    solver.add(lhs == jolts[i])

  for v in button_vars:
    solver.add(v >= 0)

  total = sum(button_vars)
  solver.minimize(total)
  solver.check()
  model = solver.model()

  return model.eval(total).as_long()

ans = 0
for machine in machines:
  ans += solve(machine)

pyperclip.copy(ans)
print(f'Copied "{ans}" to clipboard!')

File: archive/src/main/kotlin/com/grappenmaker/aoc/year25/Day10.kt
package com.grappenmaker.aoc.year25

import com.grappenmaker.aoc.simplePuzzle
import com.microsoft.z3.Expr
import com.microsoft.z3.IntSort
import com.sschr15.z3kt.*

fun main() = simplePuzzle(10, 2025) {
    partOne = inputLines.sumOf { l ->
        val parts = l.split(" ")
        val goal = parts.first().drop(1).dropLast(1).map { it == '#' }
        val buttons = parts.drop(1).dropLast(1).map { it.drop(1).dropLast(1).split(",").map(String::toInt) }

        val queue = ArrayDeque<Pair<List<Boolean>, Int>>()
        queue += goal.map { false } to 0
        val seen = hashSetOf<List<Boolean>>()

        while (queue.isNotEmpty()) {
            val (state, presses) = queue.removeLast()
            if (state == goal) return@sumOf presses

            for (b in buttons) {
                val res = state.toMutableList()
                for (i in b) res[i] = !res[i]
                if (seen.add(res)) queue.addFirst(res to (presses + 1))
            }
        }

        error("No possibility found")
    }

    z3 {
        partTwo = inputLines.sumOf { l ->
            val parts = l.split(" ")
            val buttons = parts.drop(1).dropLast(1).map { it.drop(1).dropLast(1).split(",").map(String::toInt).toSet() }
            val jolts = parts.last().drop(1).dropLast(1).split(",").map(String::toInt)
            val buttonVars = buttons.indices.map { mkIntConst("b$it") }

            mkOptimize().run {
                buttonVars.forEach { Add(it gte 0) }
                for (i in jolts.indices) Add(buttons.indices.asSequence().filter { i in buttons[it] }
                    .fold(mkInt(0)) { acc: Expr<IntSort>, curr -> mkAdd(acc, buttonVars[curr]) } eq jolts[i])

                val sum = buttonVars.fold(mkInt(0)) { acc: Expr<IntSort>, curr -> mkAdd(acc, curr) }
                MkMinimize(sum)
                Check()
                model.eval(sum, true).toLong()
            }
        }
    }
}